### ISSUE TITLE
Can you get the system latency?

### ISSUE THREAD

**Issue by tGrothmannFluffy** [RustAudio/cpal#773]
Hi there,

I am trying to get the system latency. Basically, I want to know (_roughly_) when the samples I wrote to the output are ending up at the speaker.
I know this is probably highly os and device dependent.

I found [this oboe (android) function](https://docs.rs/oboe-sys/0.4.5/oboe_sys/fn.oboe_AudioStream_calculateLatencyMillis.html) to get this value. But I was wondering if/how I could access that.

I am working on Android and iOS.

**Comment by morajabi** [comment]
I'm interested in this for macOS. Use case is for using acoustic echo cancellation works better with exact latency metrics.

**Comment by j-n-f** [comment]
Also interested in this.

From what I can tell the StreamInstant values in callbacks is just giving `block_size / sample_rate`. The value doesn't match what my driver reports (it's close, but less than the actual values).

From a quick look at ASIO and ALSA they both return this information in frame/sample counts. It would be useful to know what this looks like for the other backends.

It should be possible to use that data to provide more accurate `StreamInstant` values. Since at least ASIO + ALSA provide the latency in frames/samples, it might be nice to include that value as part of the callback info structs.

**Comment by j-n-f** [comment]
Here's what I was able to scrounge up.

My previous comment on putting this latency information at the stream config level may not always be possible. If Oboe can even provide this value, it sounds like it might be dynamically changing block size while a stream is still running.

If some backend can't provide more accurate latencies then the existing heuristic was the best you could do anyways.

This should be enough to start experimenting.

|backend|api|format|comment|
|-|-|-|-|
|ALSA|[snd_pcm_delay](https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#ga012e8b999070e72ab23514f25e7d6482)|frames/samples||
|JACK|???|???|not sure, guessing you just use ALSA to fetch this information|
|WASAPI|[IAudioClient::GetStreamLatency](https://learn.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-iaudioclient-getstreamlatency)|100ns increments||
|ASIO|ASIOGetLatencies (ASIO SDK 2.3, page 14)|samples/frames|must call after ASIOCreateBuffers <br /><br />Driver may notify host (cpal) of `kAsioLatenciesChanged` (ASIO SDK 2.3, page 9)|
|CoreAudio|[AudioObjectGetPropertyData](https://developer.apple.com/documentation/coreaudio/1422524-audioobjectgetpropertydata) (for [`kAudioDevicePropertyLatency`](https://developer.apple.com/documentation/coreaudio/kaudiodevicepropertylatency))|???|There's also some [discussion](https://lists.apple.com/archives/coreaudio-api/2010/Jan/msg00046.html) on a mailing list on how we may need to read several properties|
|Oboe|???|???|Oboe is complicated, and it seems like it might dynamically adjust block size while running. Might just have to use existing heuristic for this one. ([further reading](https://google.github.io/oboe/classoboe_1_1_latency_tuner.html))|
|Emscripten|???|???|Not actually sure how this backend works, can't find documentation|

**Comment by marcpabst** [comment]
Not sure if this is still relevant for @tGrothmannFluffy, but I've started to implemented a `latency()` method for `Stream` that returns the latency in samples/frames. @j-n-f's research was quite helpful!

**Comment by roderickvd** [comment]
Yes, you can get it through the `StreamInstant` values. Like you said, on some hosts it was close but not exact. Dropped #1129 for that.
